File: js/tilings/truncated-hexagonal.js
/**
 * Generates a Truncated Hexagonal tiling board (3.12.12).
 * This tiling consists of dodecagons and triangles.
 * Every vertex is shared by two dodecagons and one triangle.
 */
export function generateTruncatedHexagonalBoard(options) {
  const { cols, rows, tileSize: a, colorCount, rng } = options;

  // Distance between centers of dodecagons (sharing a side 'a')
  const dist = a * (2 + Math.sqrt(3));

  const tiles = [];
  const dodecaMap = new Map(); // key: "q,r", value: id
  const triangleMap = new Map(); // key: "x,y", value: id

  const getDodecaCenter = (q, r) => {
    const x = dist * (q + r / 2);
    const y = dist * (Math.sqrt(3) / 2) * r;
    return [x, y];
  };

  const idToTile = new Map();
  let idCounter = 0;

  // 1. Generate Dodecagons
  const R12 = a / (2 * Math.sin(Math.PI / 12));
  for (let r = 0; r < rows; r++) {
    const r_offset = Math.floor(r / 2);
    for (let q = -r_offset; q < cols - r_offset; q++) {
      const [cx, cy] = getDodecaCenter(q, r);
      const id = idCounter++;

      const points = [];
      for (let i = 0; i < 12; i++) {
        const angle = (30 * i + 15) * Math.PI / 180;
        points.push([cx + R12 * Math.cos(angle), cy + R12 * Math.sin(angle)]);
      }

      const tile = {
        id,
        type: 'dodecagon',
        q, r,
        colorId: Math.floor(rng() * colorCount),
        ownerId: null,
        points,
        neighbors: []
      };
      tiles.push(tile);
      dodecaMap.set(`${q},${r}`, id);
      idToTile.set(id, tile);
    }
  }

  // Helper to add or get a triangle by position
  const getTriangle = (cx, cy, angle) => {
    const key = `${Math.round(cx * 100)},${Math.round(cy * 100)}`;
    if (triangleMap.has(key)) return triangleMap.get(key);

    const id = idCounter++;
    const points = [];
    const rT = a / Math.sqrt(3);
    for (let i = 0; i < 3; i++) {
      const vAngle = angle + (i * 120) * Math.PI / 180;
      points.push([cx + rT * Math.cos(vAngle), cy + rT * Math.sin(vAngle)]);
    }

    const tile = {
      id,
      type: 'triangle',
      colorId: Math.floor(rng() * colorCount),
      ownerId: null,
      points,
      neighbors: []
    };
    tiles.push(tile);
    triangleMap.set(key, id);
    idToTile.set(id, tile);
    return id;
  };

  // 2. Build connectivity and generate triangles
  const dodecaIds = tiles.filter(t => t.type === 'dodecagon').map(t => t.id);
  const distT = dist / Math.sqrt(3);

  for (const dId of dodecaIds) {
    const dodeca = idToTile.get(dId);
    const [cx, cy] = getDodecaCenter(dodeca.q, dodeca.r);

    // Neighbors: 6 other dodecagons and 6 triangles
    const directions = [[1, 0], [1, -1], [0, -1], [-1, 0], [-1, 1], [0, 1]];
    for (const [dq, dr] of directions) {
      const neighborId = dodecaMap.get(`${dodeca.q + dq},${dodeca.r + dr}`);
      if (neighborId !== undefined) {
        const neighbor = idToTile.get(neighborId);
        if (!dodeca.neighbors.includes(neighborId)) dodeca.neighbors.push(neighborId);
        if (!neighbor.neighbors.includes(dId)) neighbor.neighbors.push(dId);
      }
    }

    for (let i = 0; i < 6; i++) {
      const angleT = (60 * i + 30) * Math.PI / 180;
      const tcx = cx + distT * Math.cos(angleT);
      const tcy = cy + distT * Math.sin(angleT);

      const tId = getTriangle(tcx, tcy, angleT + Math.PI);
      const triangle = idToTile.get(tId);

      if (!dodeca.neighbors.includes(tId)) dodeca.neighbors.push(tId);
      if (!triangle.neighbors.includes(dId)) triangle.neighbors.push(dId);
    }
  }

  // 2.5 Edge culling: remove triangles with fewer than 3 neighbors
  const tilesToKeep = [];
  const removedIds = new Set();
  for (const tile of tiles) {
    if (tile.type === 'triangle' && tile.neighbors.length < 3) {
      removedIds.add(tile.id);
    } else {
      tilesToKeep.push(tile);
    }
  }

  // Update neighbors and re-index
  const idMap = new Map();
  const finalTiles = tilesToKeep.map((tile, index) => {
    tile.neighbors = tile.neighbors.filter(nId => !removedIds.has(nId));
    idMap.set(tile.id, index);
    tile.id = index;
    return tile;
  });

  finalTiles.forEach(tile => {
    tile.neighbors = tile.neighbors.map(nId => idMap.get(nId));
  });

  // 3. Finalize
  let minX = Infinity, minY = Infinity;
  let maxX = -Infinity, maxY = -Infinity;

  for (const t of tiles) {
    for (const p of t.points) {
      minX = Math.min(minX, p[0]);
      minY = Math.min(minY, p[1]);
      maxX = Math.max(maxX, p[0]);
      maxY = Math.max(maxY, p[1]);
    }
  }

  tiles.forEach(t => {
    t.points = t.points.map(p => [p[0] - minX, p[1] - minY]);
  });

  const startTileIds = [
    dodecaMap.get(`${0},${0}`),
    dodecaMap.get(`${cols - 1},0`),
    dodecaMap.get(`${-Math.floor((rows - 1) / 2)},${rows - 1}`),
    dodecaMap.get(`${cols - 1 - Math.floor((rows - 1) / 2)},${rows - 1}`)
  ].filter(id => id !== undefined);

  return {
    version: 1,
    generator: "truncated-hexagonal",
    width: maxX - minX,
    height: maxY - minY,
    tiles: finalTiles,
    startTileIds: startTileIds.filter(id => !removedIds.has(id)).map(id => idMap.get(id))
  };
}

File: go-flood/pkg/tilings/truncated_hexagonal.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
)

func GenerateTruncatedHexagonalBoard(options Options) core.Board {
	a := options.TileSize
	dist := a * (2.0 + math.Sqrt(3.0))
	cols, rows := options.Cols, options.Rows
	colorCount := options.ColorCount
	rng := options.RNG

	tiles := []core.Tile{}
	dodecaMap := make(map[string]int)
	triangleMap := make(map[string]int)
	idToTileIdx := make(map[int]int)

	getDodecaCenter := func(q, r int) (float64, float64) {
		x := dist * (float64(q) + float64(r)/2.0)
		y := dist * (math.Sqrt(3.0) / 2.0) * float64(r)
		return x, y
	}

	idCounter := 0
	// 1. Generate Dodecagons
	R12 := a / (2.0 * math.Sin(math.Pi/12.0))
	for r := 0; r < rows; r++ {
		rOffset := int(math.Floor(float64(r) / 2.0))
		for q := -rOffset; q < cols-rOffset; q++ {
			cx, cy := getDodecaCenter(q, r)
			id := idCounter
			idCounter++

			points := make([]core.Point, 12)
			for i := 0; i < 12; i++ {
				angle := (30.0*float64(i) + 15.0) * math.Pi / 180.0
				points[i] = core.Point{cx + R12*math.Cos(angle), cy + R12*math.Sin(angle)}
			}

			tile := core.Tile{
				ID:        id,
				ColorID:   int(rng() * float64(colorCount)),
				OwnerID:   nil,
				Points:    points,
				Neighbors: []int{},
			}
			dodecaMap[fmt.Sprintf("%d,%d", q, r)] = id
			idToTileIdx[id] = len(tiles)
			tiles = append(tiles, tile)
		}
	}

	// Helper for triangles
	getTriangle := func(cx, cy, angle float64) int {
		key := fmt.Sprintf("%d,%d", int(math.Round(cx*100)), int(math.Round(cy*100)))
		if id, ok := triangleMap[key]; ok {
			return id
		}

		id := idCounter
		idCounter++
		points := make([]core.Point, 3)
		rT := a / math.Sqrt(3.0)
		for i := 0; i < 3; i++ {
			vAngle := angle + (float64(i)*120.0)*math.Pi/180.0
			points[i] = core.Point{
				cx + rT*math.Cos(vAngle),
				cy + rT*math.Sin(vAngle),
			}
		}

		tile := core.Tile{
			ID:        id,
			ColorID:   int(rng() * float64(colorCount)),
			OwnerID:   nil,
			Points:    points,
			Neighbors: []int{},
		}
		triangleMap[key] = id
		idToTileIdx[id] = len(tiles)
		tiles = append(tiles, tile)
		return id
	}

	addNeighbor := func(idx1, id2 int) {
		found := false
		for _, n := range tiles[idx1].Neighbors {
			if n == id2 {
				found = true
				break
			}
		}
		if !found {
			tiles[idx1].Neighbors = append(tiles[idx1].Neighbors, id2)
		}
	}

	distT := dist / math.Sqrt(3.0)

	// 2. Build connectivity and generate triangles
	for key, dId := range dodecaMap {
		var q, r int
		fmt.Sscanf(key, "%d,%d", &q, &r)
		cx, cy := getDodecaCenter(q, r)
		dIdx := idToTileIdx[dId]

		directions := [][2]int{{1, 0}, {1, -1}, {0, -1}, {-1, 0}, {-1, 1}, {0, 1}}
		for _, d := range directions {
			if nID, ok := dodecaMap[fmt.Sprintf("%d,%d", q+d[0], r+d[1])]; ok {
				nIdx := idToTileIdx[nID]
				addNeighbor(dIdx, nID)
				addNeighbor(nIdx, dId)
			}
		}

		for i := 0; i < 6; i++ {
			angleT := (60.0*float64(i) + 30.0) * math.Pi / 180.0
			tcx := cx + distT*math.Cos(angleT)
			tcy := cy + distT*math.Sin(angleT)
			tId := getTriangle(tcx, tcy, angleT+math.Pi)
			tIdx := idToTileIdx[tId]

			addNeighbor(dIdx, tId)
			addNeighbor(tIdx, dId)
		}
	}

	// 2.5 Edge culling: remove triangles with fewer than 3 neighbors
	removedIds := make(map[int]bool)
	finalTiles := []core.Tile{}
	for _, t := range tiles {
		isTriangle := false
		for _, id := range triangleMap {
			if t.ID == id {
				isTriangle = true
				break
			}
		}
		if isTriangle && len(t.Neighbors) < 3 {
			removedIds[t.ID] = true
		} else {
			finalTiles = append(finalTiles, t)
		}
	}

	// Re-map IDs and update neighbors
	idMap := make(map[int]int)
	for i := range finalTiles {
		oldId := finalTiles[i].ID
		finalTiles[i].ID = i
		idMap[oldId] = i
	}

	for i := range finalTiles {
		newNeighbors := []int{}
		for _, nId := range finalTiles[i].Neighbors {
			if !removedIds[nId] {
				newNeighbors = append(newNeighbors, idMap[nId])
			}
		}
		finalTiles[i].Neighbors = newNeighbors
	}

	// 3. Finalize
	minX, minY := math.MaxFloat64, math.MaxFloat64
	maxX, maxY := -math.MaxFloat64, -math.MaxFloat64
	for _, t := range tiles {
		for _, p := range t.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}

	for i := range tiles {
		for j := range tiles[i].Points {
			tiles[i].Points[j][0] -= minX
			tiles[i].Points[j][1] -= minY
		}
	}

	startTileIds := []int{}
	corners := []struct{ q, r int }{
		{0, 0},
		{cols - 1, 0},
		{-int(math.Floor(float64(rows-1) / 2.0)), rows - 1},
		{cols - 1 - int(math.Floor(float64(rows-1)/2.0)), rows - 1},
	}
	for _, c := range corners {
		if id, ok := dodecaMap[fmt.Sprintf("%d,%d", c.q, c.r)]; ok {
			startTileIds = append(startTileIds, id)
		}
	}

	finalStartTileIds := []int{}
	for _, id := range startTileIds {
		if !removedIds[id] {
			finalStartTileIds = append(finalStartTileIds, idMap[id])
		}
	}

	return core.Board{
		Version:      1,
		Generator:    "truncated-hexagonal",
		Width:        maxX - minX,
		Height:       maxY - minY,
		Cols:         cols,
		Rows:         rows,
		Tiles:        finalTiles,
		StartTileIds: finalStartTileIds,
	}
}
